### PR TITLE
Support Channelz Pagination

### PR DESCRIPTION
**Pull request by ncteisen**
100 was chosen because it seemed reasonable. We can bikeshed it.

The test has too much hardcoding, but I didn't really want to expose kPaginationLimit.

Tracking #15340 and #15986

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                         FILE SIZE
 ++++++++++++++ GROWING                                                           ++++++++++++++
  [ = ]       0 [None]                                                               +928  +0.0%
  +4.1%    +128 src/core/lib/channel/channelz_registry.cc                            +128  +4.1%
       +11%     +73 grpc_core::channelz::ChannelzRegistry::InternalGetServers             +73   +11%
       +10%     +70 grpc_core::channelz::ChannelzRegistry::InternalGetTopChannels         +70   +10%

  +0.0%    +128 TOTAL                                                             +1.03Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,922      Total (=)      1,965,922

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,008      Total (>)     10,908,998

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/channel/channelz_registry.cc [hunk 1]
```diff
@@ -133,10 +136,18 @@ char* ChannelzRegistry::InternalGetTopChannels(intptr_t start_channel_id) {
   // start_channel_id=0, which signifies "give me everything." Hence this
   // funky looking line below.
   size_t start_idx = start_channel_id == 0 ? 0 : start_channel_id - 1;
+  bool more_to_come = false;
```
Comment (markdroth) [inline src/core/lib/channel/channelz_registry.cc]: Suggest renaming this to something like `reached_pagination_limit`.

File: src/core/lib/channel/channelz_registry.cc [hunk 2]
```diff
@@ -169,10 +180,18 @@ char* ChannelzRegistry::InternalGetServers(intptr_t start_server_id) {
   // reserved). However, we want to support requests coming in with
   // start_server_id=0, which signifies "give me everything."
   size_t start_idx = start_server_id == 0 ? 0 : start_server_id - 1;
+  bool more_to_come = false;
```
Comment (markdroth) [inline src/core/lib/channel/channelz_registry.cc]: Same here.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                         FILE SIZE
 ++++++++++++++ GROWING                                                           ++++++++++++++
  [ = ]       0 [None]                                                               +944  +0.0%
  +4.1%    +128 src/core/lib/channel/channelz_registry.cc                            +128  +4.1%
       +11%     +73 grpc_core::channelz::ChannelzRegistry::InternalGetServers             +73   +11%
       +10%     +70 grpc_core::channelz::ChannelzRegistry::InternalGetTopChannels         +70   +10%

  +0.0%    +128 TOTAL                                                             +1.05Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by carl-mastrangelo** [comment]
@markdroth Ack.  I think max results was in the original design, but now I can't remember why it was removed.

File: src/core/lib/channel/channelz_registry.cc [hunk 3]
```diff
@@ -133,10 +136,18 @@ char* ChannelzRegistry::InternalGetTopChannels(intptr_t start_channel_id) {
...
+  bool reached_pagination_limit = false;
   for (size_t i = start_idx; i < entities_.size(); ++i) {
     if (entities_[i] != nullptr &&
         entities_[i]->type() ==
             grpc_core::channelz::BaseNode::EntityType::kTopLevelChannel) {
+      // check if we are over pagination limit to determine if we need to set
+      // the "end" element. If we don't go through this block, we know that
+      // when the loop terminates, we have <= to kPaginationLimit.
+      if (top_level_channels.size() == kPaginationLimit) {
+        reached_pagination_limit = true;
```
Comment (carl-mastrangelo) [inline src/core/lib/channel/channelz_registry.cc]: Is this correct?  If there are 101 channelz, won't this erroneously set the "end" field down below?  Or end mean something else?

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,218      Total (=)      1,965,218

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,026      Total (<)     10,909,030

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/channel/channelz_registry.cc (hunk 3 above)
Comment (ncteisen) [reply src/core/lib/channel/channelz_registry.cc]: If there are 101 channels then reached_pagination_limit=true just before we break, so we will NOT set "end' below. so end will be false

**Pull request merged**